fix the -runtime-variant option for bytecode

damiendoligez opened the pull request:
The `-runtime-variant` option has been broken since 4.03.0 because the `camlheaderd` and `camlheaderi` files are installed under the wrong names. The first commit fixes this problem.

The failure is silent and a bytecode file is produced that has the executable flag but doesn't work because the header is missing. The second commit adds some error handling for this case (in particular when the user specifies a runtime variant that doesn't exist).

Note: the first bug was introduced by myself in commit 0225ca01e39289ce1801fb09cd011cdbfb542b8d

damiendoligez committed:
Fix the installation of the camlheader[di] files. They were installed under
the wrong names, which made -runtime-variant fail for bytecode.

damiendoligez committed:
Add error handling when finding and copying the camlheader file.

damiendoligez commented:
See also #2309 and #6199.

dra27 commented on Changes:
```diff
@@ -134,10 +134,14 @@ Working version
   declaration
   (Florian Angeletti, review by Jacques Garrigue)
 
+- GPR#8537: fix the -runtime-variant option for byte-code
```
```suggestion
- GPR#8537: fix the -runtime-variant option for bytecode
```
"byte-code" seems to have gone the way of "O'Caml" 🙂

dra27 commented on bytecomp/bytelink.ml:
```diff
@@ -695,7 +698,9 @@ let report_error ppf = function
         Location.print_filename file
   | Required_module_unavailable s ->
-      fprintf ppf "Required module `%s' is unavailable" s
+      fprintf ppf "Required module '%s' is unavailable" s
```
Why the elimination of the back-tick here? FWIW, I'd prefer to see other instances of `'%s'` changed to `` `%s'`` rather than this way around?

dbuenzli replied:
@dra27 some people from your faculty [may disagree](https://www.cl.cam.ac.uk/~mgk25/ucs/quotes.html) ;-)

dra27 replied:
Quite possibly - but let's remove them wholesale, rather than piecemeal with other changes 🙂

dra27 replied:
(I'd also prefer them either to be post-processed to the correct Unicode character before going to the console, or for us finally to allow the correct Unicode characters directly in the message, both of which would keep Markus happy!)

dra27 commented on bytecomp/bytelink.ml:
```diff
@@ -695,7 +698,9 @@ let report_error ppf = function
-      fprintf ppf "Required module `%s' is unavailable" s
+      fprintf ppf "Required module '%s' is unavailable" s
+  | Camlheader (msg, header) ->
+      fprintf ppf "System error while copying file %s : %s" header msg
```
```suggestion
      fprintf ppf "System error while copying file %s: %s" header msg
```

dra27 reviewed (commented):
Also related to #2267. I appear to be in a very nitty mood this morning with the other comments.

dra27 commented:
I have a nagging concern that the flexlink bootstrap relies on the "silent fail" behaviour when the header is missing - I just want to check what's going on there before this is merged (I'm suspicious that in bytecode-only mode, it may result in a corrupt `flexlink.exe` being installed).

dra27 commented:
Just to confirm - this PR will break the flexdll bootstrap build the next time someone commits a bootstrap. Do you mind if I push fixes directly?

dra27 committed:
Missing -nostdlib options in flexdll bootstrap

If OCaml was already installed, camlheader would be used from the
installation when building the bytecode flexdll/flexlink.exe

dra27 committed:
Always build flexdll/flexlink.exe with camlheader

dra27 commented:
I've rebased and added commits to fix the flexdll bootstrap in [my fork](https://github.com/dra27/ocaml/tree/runtime-variant-fixes)

damiendoligez commented:
@dra27 I've applied all your suggestions and integrated your flexdll fixes (tested by bootstrapping flexdll on mingw64).

I've also moved the Change entry to 4.09, as I intend to cherry-pick this patch.

dra27 commented:
@damiendoligez - squash or merge, and still good for 4.09?

gasche commented:
This is a bugfix so it should be good for 4.09.
(I have a marked preference for having merge commits. In this case the history is not great, but we don't want an extra roundtrip with the contributor, and it's not so bad, so I would just Merge directly.)

dra27 commented:
FWIW, I just squashed the Changes bit - the flexdll and other changes are a bit subtle, so seem worth retaining cleanly for bisecting

dra27 commented:
I'll merge and cherry-pick once CI passes

dra27 commented:
4.09: 617153cec62311a03d2dd4e83ff7ffd6bd0de8f1

damiendoligez commented:
Thanks!
